GenEra Conda Recipes Submission

AnupamGautam opened the pull request:
Dear Administrator,

Please find attached the Conda recipes for GenEra. Let us know if any further modifications are required. We look forward to your feedback.

Best regards,
Anupam

aliciaaevans commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,34 @@
+
+requirements:
+  host:
+    - python =3.7.12
```
Bioconda only builds Python 3.8 and up.

AnupamGautam replied:
Thank you will update it.

aliciaaevans commented on recipes/genera/build.sh:
```diff
@@ -0,0 +1,30 @@
+chmod +x genEra Erassignment hmmEra tree2ncbitax FASTSTEP3R test_installation.sh
+
+# Clone the secondary repository and move the main script
+git clone https://github.com/caraweisman/abSENSE.git
```
Don't use git clone. This is a separate tool, so please add it as a separate recipe.

AnupamGautam replied:
I’ve submitted a request on their GitHub for a release, but if they don’t create one, would it be possible to proceed with git clone despite the recommendation to avoid it?

martin-g replied:
You can use 'git_url: ...' and 'git_rev: commit_sha' for it

AnupamGautam replied:
Dear @aliciaaevans and @martin-g,

I have incorporated the suggested changes. Please let me know if any further adjustments are needed.

Best regards,  
Anupam

AnupamGautam replied:
Dear @aliciaaevans and @martin-g,

The author of abSENSE has made a new release, so I’ve added it. However, I’m not certain if they would like a Conda package for it. I’ve reached out to confirm, and if they are interested, I can create a PR for the Conda package as well.

Best regards,
Anupam

aliciaaevans commented on recipes/genera/build.sh:
```diff
@@ -0,0 +1,30 @@
+# Ensure all main scripts are executable
+chmod +x genEra Erassignment hmmEra tree2ncbitax FASTSTEP3R test_installation.sh
+
+# Clone the secondary repository and move the main script
+git clone https://github.com/caraweisman/abSENSE.git
+mv abSENSE/Run_abSENSE.py .
+chmod +x Run_abSENSE.py
```
You are right that the git clone command should be removed, but instead of adding the missing dependency as a source, a separate recipe should be created for the dependency in a new PR and then the dependencies updated in this recipe.

AnupamGautam replied:
Hi @aliciaaevans, thank you for the review. This particular script doesn’t have a release version as I believe it’s part of a publication’s codebase. I’ve submitted a request on their GitHub for a release, but if they don’t create one, would it be possible to proceed with `git clone` despite the recommendation to avoid it?

AnupamGautam replied:
I have added the required recipe for abSENSE (it has been merged https://anaconda.org/bioconda/absense), removed the corresponding section from GenEra, and made the necessary adjustments

AnupamGautam commented:
@BiocondaBot please add label

AnupamGautam commented:
Dear @aliciaaevans and @martin-g,

Could you please review the updated recipe and assist with the merge? Apologies for interrupting during a busy schedule, and thank you in advance for your help!

Best,
Anupam

bgruening commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+
+about:
+  home: https://github.com/josuebarrera/GenEra
+  license: "GPL-3.0"
```
Please use SDPX identifiers

AnupamGautam replied:
added

bgruening commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+  host:
+    - python =3.8.0
+  run:
+    - python =3.8.0
```
please do not spin that strict, this is not how conda works. Strict pins can be used in environments.yml files, but packages should specify pins when they are really incompatible.

bgruening commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,42 @@
+    folder: abSENSE
+
+build:
+  number: 0
```
This package should be noarch generic I think

AnupamGautam replied:
Added.

mfansler commented:
My recommendation is to use the newer `r-phytools` if possible. I almost never recommend for *recipes*<sup>**1**</sup> to use exact (`=`) specifications for dependencies. Only in cases with strong reasons should that be done.

Besides, Bioconda has adopted a `run_exports` strategy that serves to safeguard against using versions that deviate too far from those used to build the package. That doesn't, however, extend to Conda Forge packages, but one can effectively mimic it by providing bounds.

Otherwise, a proper solution to using either (1) older Bioconda packages from feedstocks that migrated to Conda Forge or (2) new Bioconda packages from feedstocks that migrated from Conda Forge has not been agreed upon. There is [a proposal](https://github.com/conda-forge/cfep/pull/52) to solve this from the Conda Forge side, but it still needs more feedback from Bioconda devs. Alternatively, Bioconda could allow recipes to use flexible solves.<sup>**2**</sup>

---

<sub>**[1]**: Environments on the other hand that go into pipelines should be exact, at least up to (semantically) minor versions.</sub>
<sub>**[2]**: My hot take on this is that `strict` channel priority is outdated and `flexible` is sufficiently reliable.</sub>

AnupamGautam commented:
Hi @mfansler,

Thank you for your help. As suggested, I am working on loosening the version requirement. I will test it locally, and if it works, I will update the recipe accordingly.

Thank you again for your support.

Best regards,
Anupam

mfansler commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+
+test:
+  commands:
+    - genEra -h 2>&1 > /dev/null
```
It would be better to see the results in the logs. Also, there are 7 executables being moved into `CONDABIN` - they should also be tested.

```suggestion
    - genEra -h
```

mfansler commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+source:
+  - url: https://github.com/josuebarrera/GenEra/archive/refs/tags/v1.4.2.tar.gz
+    sha256: acc49698b68151550b825c73f1f03dd268da2c73ecc7ac4c55eccb8bec5b1a54
+  - url: https://github.com/caraweisman/abSENSE/archive/refs/tags/v1.0.0.tar.gz
+    sha256: a99090c10a66afae1fbad38d79d228199e9fbdae5173e71ad5d84b4c958dda54
+    folder: abSENSE
```
Is there a reason this is not a separate recipe?

AnupamGautam replied:
Dear @mfansler,  

I have implemented your suggestions and added a new test as well. Regarding the tool *abSENSE*, I have reached out to the author to confirm their interest in having a Conda package. If they express interest, I can create a PR for the package.  

For now, we can proceed like this, and if they author want I can create a PR later ,and update this package also. Could you please approve the workflow so that it can start? Thank you so much for your help!  

Best regards,  
Anupam

mfansler replied:
As far as I see, abSENSE has no licensing declared, which is prohibitive to bundling it for Bioconda. Please ask the author to declare a license and include that in a release.

Once there is a license it should be moved to a separate package here. Note that the author does not have to be a maintainer here.

mfansler replied:
Thanks for the additional information!

While the article itself is CC-BY-4.0, a copy of the code does not appear deposited with the article and thus it is not covered. If they want to extend exactly the same license, they could put a copy of [the CC-BY-4.0 text](https://creativecommons.org/licenses/by/4.0/legalcode.txt) in the repository somewhere.

Otherwise, unfortunately "open-source" doesn't translate to permission to redistribute.

AnupamGautam replied:
Here’s a polished version of your message:  

---

**Subject:** Update on Package and Request for Workflow Initiation  

Dear @mfansler,  

The authors have added a license file to their repository and released a new version. I’ve updated the `meta.yaml` file to include this latest release. While they may still need time to decide whether they want an official Conda package, I can create one for them later if requested and make the necessary updates.  

For now, we can proceed with the current setup. Could you help initiate the workflow and assist in merging this?  

Thank you for your continued support!  

Best regards,  
Anupam

AnupamGautam replied:
Dear @mfansler,

Thank you so much for your help—I really appreciate it. Let's wait and see what others have to say. If needed, I can also include this script in our package or create a dedicated Conda package for it.

Best regards,
Anupam

mfansler commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+package:
+  name: genera
+  version: "1.4.2"
+
+source:
+  - url: https://github.com/josuebarrera/GenEra/archive/refs/tags/v1.4.2.tar.gz
```
Version could be a variable. Here's canonical boilerplate for this:

```suggestion
{% set version = '1.4.2' %}

package:
   name: genera
   version: {{ version|replace('-', '_') }}

source:
  - url: https://github.com/josuebarrera/GenEra/archive/refs/tags/v{{ version }}.tar.gz
```

AnupamGautam commented:
Dear @mfansler and @aliciaaevans,  

I have added the required recipe for **abSENSE** (it has been merged https://anaconda.org/bioconda/absense), removed the corresponding section from **GenEra**, and made the necessary adjustments. The updated recipe has passed all checks.  

I believe this issue can now be resolved and merged. Please let me know if further changes are needed.  

Best regards,  
Anupam

mfansler commented on recipes/genera/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+    - {{ pin_subpackage('genera', max_pin="x.x") }}
+
+requirements:
+  host:
+    - python >=3.8,<3.9
+  run:
+    - python >=3.8,<3.9
+    - diamond >=2.1.10
+    - mmseqs2 =14.7e284
+    - foldseek =3.915ef7d
+    - scipy =1.7.3
+    - r-seqinr =4.2_16
+    - r-optparse =1.7.3
+    - r-bio3d =2.4_3
+    - r-tidyverse =1.3.2
+    - r-phytools >=0.6_99
+    - orthofinder =2.5.5
+    - ncbitax2lin =2.3.2
```
I agree with the bot here. Are the exact specifications really necessary?

The `run_exports` of other Bioconda packages will add upper bounds that help constrain co-installation with too new versions in the future.

Also, [Python 3.8 has reached EOL](https://devguide.python.org/versions/#python-release-cycle), so Python 3.9 would make more sense as a minimum.

The `pip` and `setuptools` suggested by the bot can be ignored - they aren't relevant since this only copies scripts.

AnupamGautam commented:
Dear @mfansler,  

I have adjusted the flexibility wherever possible. However, Python 3.8 (or versions below 3.9) is currently required as **abSENSE** is not compatible with Python >=3.9. Until the author provides an update, we will need to maintain this version requirement. Once they implement compatibility, I will update the recipe accordingly.  

Regarding other dependencies, some flags are only available in specific versions, which necessitates retaining version-specific requirements for now.  

Best regards,  
Anupam

mfansler reviewed (approved):
Okay, let's merge.

---

**Note**

> I have adjusted the flexibility wherever possible. However, Python 3.8 (or versions below 3.9) is currently required as **abSENSE** is not compatible with Python >=3.9. Until the author provides an update, we will need to maintain this version requirement. Once they implement compatibility, I will update the recipe accordingly.

Dependencies do not usually dictate the Python compatibility of a package. While that might restrict the environments in which it can be installed, the Python compatibility should be intrinsic to the package code (e.g., what syntactic features it uses). In this line, were an `abSENSE` that is Python v3.9+ released, one would immediately have the ability to co-install with it, rather than having to build and upload new artifacts.

However, I won't insist on this.

AnupamGautam commented:
Dear @mfansler,

Thank you so much for your approval and support; I truly appreciate it. I believe I’ll need to wait for @aliciaaevans before the changes can be merged.

Best regards,  
Anupam

aliciaaevans commented:
@BiocondaBot please fetch artifacts
